tensorflow/compiler/xla/service/pattern_matcher_test.cc

```diff
@@ -1037,6 +1040,62 @@ TEST_F(PatternMatcherTest, OneUseAndOneUser) {
             "in p0 = f32[] parameter(0)");
 }
 
+TEST_F(PatternMatcherTest, MatchSingleUserOperands) {
+  auto param =
+      HloInstruction::CreateParameter(0, ShapeUtil::MakeShape(F32, {}), "p");
+
+  auto reshape =
+      SetName("reshape", HloInstruction::CreateReshape(
+                             ShapeUtil::MakeShape(F32, {1}), param.get()));
+  EXPECT_TRUE(MatchSingleUserOperands(reshape.get(), m::Reshape(m::Op())));
+
+  auto bitcast =
+      SetName("bitcast", HloInstruction::CreateBitcast(
+                             ShapeUtil::MakeShape(F32, {1}), param.get()));
+  EXPECT_TRUE(MatchSingleUserOperands(param.get(), m::Op()));
+  EXPECT_TRUE(MatchSingleUserOperands(bitcast.get(), m::Bitcast()));
+  EXPECT_FALSE(MatchSingleUserOperands(bitcast.get(), m::Bitcast(m::Op())));
+  EXPECT_EQ(Explanation(bitcast.get(), m::Bitcast(m::Op()), true),
+            "Operand 0 of HloInstruction has 2 users. Expected 1.\nin bitcast "
+            "= f32[1]{0} bitcast(f32[] p)");
+
+  auto param0 =
+      HloInstruction::CreateParameter(0, ShapeUtil::MakeShape(F32, {}), "p0");
+  auto param1 =
+      HloInstruction::CreateParameter(0, ShapeUtil::MakeShape(F32, {}), "p1");
+
+  auto add = SetName("add", HloInstruction::CreateBinary(
+                                ShapeUtil::MakeShape(F32, {}), HloOpcode::kAdd,
+                                param0.get(), param0.get()));
+  EXPECT_TRUE(MatchSingleUserOperands(add.get(), m::Add(m::Op(), m::Op())));
+
+  auto mul =
+      SetName("mul", HloInstruction::CreateBinary(ShapeUtil::MakeShape(F32, {}),
+                                                  HloOpcode::kMultiply,
+                                                  param1.get(), param0.get()));
+  EXPECT_TRUE(MatchSingleUserOperands(mul.get(), m::Multiply()));
```

Comment:
Does this pass? What users does multiply itself have?

Reply:
This passes because the top-level operation in the prescribed pattern can have an arbitrary number of users, including zero. The requirement of a single user only applies to the intermediates, i.e. the operands of the top-level operation (and their respective operands).